static/app/components/tours/components.tsx

```diff
@@ -0,0 +1,245 @@
+
+interface TourContextProviderProps<T extends TourEnumType> {
+  /**
+   * The children of the tour context provider.
+   * All children of this component will be blurred when the tour is active.
+   * Only the active tour element and overlay will be discernible.
+   */
+  children: React.ReactNode;
+  /**
+   * Whether the tour can be accessed by the user
+   */
+  isAvailable: TourState<T>['isAvailable'];
+  /**
+   * The ordered list of Step IDs
+   */
+  orderedStepIds: TourState<T>['orderedStepIds'];
+  /**
+   * The React context (from createContext) containing the provider for the tour.
+   * The value for this prop comes from useTourReducer, to avoid extra steps.
+   */
+  tourContext: React.Context<TourContextType<T>>;
+}
+
+export function TourContextProvider<T extends TourEnumType>({
+  children,
+  isAvailable,
+  tourContext,
+  orderedStepIds,
+}: TourContextProviderProps<T>) {
+  const tourContextValue = useTourReducer<T>({
+    isAvailable,
+    orderedStepIds,
+    currentStep: null,
+  });
+  const {dispatch, currentStep} = tourContextValue;
+  const isTourActive = currentStep !== null;
+
+  const tourHotkeys = useMemo(() => {
+    return [
+      {match: 'Escape', callback: () => dispatch({type: 'END_TOUR'})},
+      {match: ['left', 'h'], callback: () => dispatch({type: 'PREVIOUS_STEP'})},
+      {match: ['right', 'l'], callback: () => dispatch({type: 'NEXT_STEP'})},
+    ];
+  }, [dispatch]);
+
+  useHotkeys(tourHotkeys);
+
+  return (
+    <tourContext.Provider value={tourContextValue}>
+      {isTourActive && <BlurWindow />}
+      {children}
+    </tourContext.Provider>
+  );
+}
+
+export interface TourElementProps<T extends TourEnumType>
+  extends Partial<UseOverlayProps> {
+  /**
+   * The content being focused during the tour.
+   */
+  children: React.ReactNode;
+  /**
+   * The description of the tour step.
+   */
+  description: React.ReactNode;
+  /**
+   * The unique identifier of the tour step.
+   */
+  id: TourStep<T>['id'];
+  /**
+   * The title of the tour step.
+   */
+  title: React.ReactNode;
+  /**
+   * The relevant tour context
+   */
+  tourContext: TourContextType<T>;
+  /**
+   * The className of the wrapper element.
+   */
+  className?: string;
+}
+
+export function TourElement<T extends TourEnumType>({
+  children,
+  id,
+  title,
+  description,
+  tourContext,
+  position,
+  className,
+}: TourElementProps<T>) {
+  const theme = useTheme();
+
+  const {currentStep, dispatch, orderedStepIds, registerStep} = tourContext;
+  const stepCount = currentStep ? orderedStepIds.indexOf(id) + 1 : 0;
+  const stepTotal = orderedStepIds.length;
+  const hasPreviousStep = stepCount > 1;
+  const hasNextStep = stepCount < stepTotal;
+  const isOpen = currentStep?.id === id;
+
+  const {triggerProps, triggerRef, overlayProps} = useOverlay({
+    isOpen,
+    position,
+    disableTrigger: true,
+  });
+  const element = triggerRef.current;
+
+  useEffect(() => {
+    registerStep({id, element});
```

Comment:
I think `registerStep` should return a function that un-registers the step as well, so if it un-renders it correctly is removed from being registered in the context

```tsx
useEffect(() => registerStep({id, element: triggerRef.current}), [id, registerStep]);
```

Alternative I would abstract this into it's own hook exported from tourContext

```tsx
useRegisterTourStep({id, element: triggerRef.current});
```

Reply:
though you probably don't _really_ need to do that abstraction since this is already inside the common tour component.

Reply:
I’m not really sure I understand why we’d need this `registry` ? It goes a bit against the react data flow that usually goes from parent -> child, and writing to a `ref` means we’re rendering components based on ref values, which is slightly against the rules of react.

From what I can see, we have a provider that gets a list of `orderedStepIds` passed. I guess it’s expected that all elements with those step ids will eventually render? Otherwise, the `stepCount` calculation in the `TourElement` won’t be correct.

A simplification would just increase the `currentStepId` when `Next` is clicked by taking the next element in the `orderedStepIds` unconditionally.

Or am I missing a scenario where we have something in side the `orderStepIds` that doesn’t render 🤔?

Reply:
The registry is for that scenario, where the components needed for the tour do not render, or take long to render. As written, the goal is to only allow tours to be started once all the steps have been registered (meaning they've been rendered in some way). We're using a ref to avoid re-rendering every other step when one step registers for the first time.